File: nim/src/call_coding_clis/help.nim
import std/os
import std/osproc
import std/strutils
import std/streams
import std/tables
import call_coding_clis/parser

type
  RunnerEntry = tuple[name: string, alias: string]

const CANONICAL_RUNNERS: seq[RunnerEntry] = @[
  ("opencode", "oc"),
  ("claude", "cc"),
  ("kimi", "k"),
  ("codex", "rc"),
  ("crush", "cr"),
]

const HELP_TEXT = """ccc — call coding CLIs

Usage:
  ccc [runner] [+thinking] [:provider:model] [@name] "<Prompt>"
  ccc --help
  ccc -h

Slots (in order):
  runner        Select which coding CLI to use (default: oc)
                opencode (oc), claude (cc), kimi (k), codex (rc), crush (cr)
  +thinking     Set thinking level: +0 (off) through +4 (max)
  :provider:model  Override provider and model
  @name         Use a named preset from config; if no preset exists, treat it as an agent

Examples:
  ccc "Fix the failing tests"
  ccc oc "Refactor auth module"
  ccc cc +2 :anthropic:claude-sonnet-4-20250514 "Add tests"
  ccc k +4 "Debug the parser"
  ccc @reviewer "Audit the API boundary"
  ccc codex "Write a unit test"

Config:
  ~/.config/ccc/config.toml  — default runner, presets, abbreviations
"""

proc getVersion(binary: string): string =
  try:
    var p = startProcess(command = binary, args = @["--version"], options = {poUsePath})
    let outStrm = p.outputStream()
    let output = outStrm.readAll()
    let exitCode = p.waitForExit(3000)
    if exitCode == 0 and output.strip().len > 0:
      result = output.strip().splitLines()[0]
    p.close()
  except OSError, ValueError, IOError:
    discard

proc runnerChecklist(): string =
  let registry = runnerRegistry()
  var lines: seq[string] = @["Runners:"]
  for (name, _) in CANONICAL_RUNNERS:
    let info = registry.getOrDefault(name)
    let binary = if info.binary.len > 0: info.binary else: name
    let found = findExe(binary).len > 0
    if found:
      let ver = getVersion(binary)
      let tag = if ver.len > 0: ver else: "found"
      lines.add("  [+] " & align(name, 10) & " (" & binary & ")  " & tag)
    else:
      lines.add("  [-] " & align(name, 10) & " (" & binary & ")  not found")
  return lines.join("\n")

proc helpText*(): string =
  result = HELP_TEXT

proc usageText*(): string =
  result = "usage: ccc [runner] [+thinking] [:provider:model] [@name] \"<Prompt>\""

proc helpOutput*(): string =
  result = strip(helpText(), leading = false, trailing = true, chars = {'\n'}) & "\n\n" & runnerChecklist()

proc usageOutput*(): string =
  result = usageText() & "\n" & runnerChecklist()

proc printHelp*() =
  echo helpOutput()

proc printUsage*() =
  write stderr, usageOutput() & "\n"

File: fsharp/src/Library/Help.fs
namespace CallCodingClis

module Help =

    open System
    open System.Diagnostics

    let private canonicalRunners =
        [ ("opencode", "oc")
          ("claude", "cc")
          ("kimi", "k")
          ("codex", "rc")
          ("crush", "cr") ]

    let helpText =
        """ccc — call coding CLIs

Usage:
  ccc [runner] [+thinking] [:provider:model] [@alias] "<Prompt>"
  ccc --help
  ccc -h

Slots (in order):
  runner        Select which coding CLI to use (default: oc)
                opencode (oc), claude (cc), kimi (k), codex (rc), crush (cr)
  +thinking     Set thinking level: +0 (off) through +4 (max)
  :provider:model  Override provider and model
  @alias        Use a named preset from config

Examples:
  ccc "Fix the failing tests"
  ccc oc "Refactor auth module"
  ccc cc +2 :anthropic:claude-sonnet-4-20250514 "Add tests"
  ccc k +4 "Debug the parser"
  ccc codex "Write a unit test"

Config:
  ~/.config/ccc/config.toml  — default runner, aliases, abbreviations
"""

    let private getVersion (binary: string) : string =
        try
            let psi = ProcessStartInfo(binary, "--version")
            psi.UseShellExecute <- false
            psi.RedirectStandardOutput <- true
            psi.RedirectStandardError <- false
            let proc = Process.Start(psi)
            if proc.WaitForExit(3000) && proc.ExitCode = 0 then
                let output = proc.StandardOutput.ReadToEnd().Trim()
                if not (String.IsNullOrEmpty(output)) then
                    output.Split('\n').[0]
                else
                    ""
            else
                ""
        with
        | _ -> ""

    let private which (name: string) : bool =
        try
            let psi = ProcessStartInfo("which", name)
            psi.UseShellExecute <- false
            psi.RedirectStandardOutput <- true
            psi.RedirectStandardError <- true
            let proc = Process.Start(psi)
            proc.WaitForExit() |> ignore
            proc.ExitCode = 0
        with
        | _ -> false

    let runnerChecklist () : string =
        let lines = ResizeArray<string>()
        lines.Add("Runners:")
        for name, alias in canonicalRunners do
            let binary =
                match Parser.runnerRegistry.TryFind(name) with
                | Some info -> info.Binary
                | None -> name
            let found = which binary
            if found then
                let version = getVersion binary
                let tag = if not (String.IsNullOrEmpty(version)) then version else "found"
                lines.Add(sprintf "  [+] %-10s (%s)  %s" name binary tag)
            else
                lines.Add(sprintf "  [-] %-10s (%s)  not found" name binary)
        String.Join("\n", lines)

    let printHelp () : unit =
        printfn "%s" (helpText.TrimEnd())
        printfn ""
        printfn "%s" (runnerChecklist ())

    let printUsage () : unit =
        eprintfn "usage: ccc [runner] [+thinking] [:provider:model] [@alias] \"<Prompt>\""
        eprintfn "%s" (runnerChecklist ())
